### ISSUE TITLE
Struct field names are checked for duplication globally

### ISSUE THREAD

**Issue by sockmaster27** [flix/flix#8405]
The program
```flix
struct A[r: Region] {
    x: Int32
}

struct B[r: Region] {
    x: String
}
```
gives the error
```
❌ -- Name Error -------------------------------------------------- file:///c%3A/Users/Holger/Documents/Flix/test/Main.flix

>> Duplicate definition of '€x'.

2 |     x: Int32
        ^
        the first definition was here.

6 |     x: String
        ^
        the second definition was here.
```
on top of 56 other resolution errors from inside the standard library.

**Comment by magnus-madsen** [comment]
@dghosef I suppose this should not happen?

**Comment by dghosef** [comment]
I'm not sure about the 56 other resolution errors, but this is a legitimate error. Otherwise, if you had
`expr€x`, you wouldn't be able to determine if `expr` was `A` or `B`

**Comment by magnus-madsen** [comment]
> I'm not sure about the 56 other resolution errors, but this is a legitimate error. Otherwise, if you had `expr€x`, you wouldn't be able to determine if `expr` was `A` or `B`

But I think that `x` is only visible inside the companion module, i.e. A or B. If you write `expr€x` at the top then there is no such `x`.

**Comment by dghosef** [comment]
What should the behavior of
```
struct A[r: Region] {
    x: Int32
}

struct B[r: Region] {
    x: String
}
def foo(s: a): Unit = {
    s€x
}
```
be?

Right now, you're able to access `x` from the global namespace which feels correct(otherwise, for example, the stack example won't work).

**Comment by dghosef** [comment]
no they're not in the example

**Comment by magnus-madsen** [comment]
I see. But! They definitely should be. I think the point is that the internals of a struct is only visible inside its module. 

That's where you write public and private functions that use the struct.

This design is perhaps a bit unusual, but it is consistent with the rest of Flix.

**Comment by magnus-madsen** [comment]
(Also, it avoids the above problem)

**Comment by dghosef** [comment]
Ok, so to clarify, the following should fail to compile:
```
struct S[r] {
    x: Int32
}
def foo(S[r]): Int32 \ r = s€x
```
But this should compile:
```
struct S[r] {
    x: Int32
}
mod S {
    def foo(S[r]): Int32 \ r = s€x
}
```
And this:
```

mod S {
    struct S[r] {
        x: Int32
    }
    def foo(S[r]): Int32 \ r = s€x
}
```

**Comment by magnus-madsen** [comment]
The last example should not compile because the companion module is (strangely) called `S.S` and the access is not inside `S.S` only `S`.

**Comment by magnus-madsen** [comment]
@mlutze @JonathanStarup Do you agree with the post above?

**Comment by sockmaster27** [comment]
Why does this
> ```
> struct A[r: Region] {
>     x: Int32
> }
> 
> struct B[r: Region] {
>     x: String
> }
> def foo(s: a): Unit = {
>     s€x
> }
> ```
not just give an error since `a` isn't restricted to a type with an `x` field?

**Comment by dghosef** [comment]
Because currently struct field stuff is done before typing

**Comment by sockmaster27** [comment]
that sounds wild

**Comment by JonathanStarup** [comment]
> @mlutze @JonathanStarup Do you agree with the post above?

i agree yes

**Comment by dghosef** [comment]
Ok, so basically, we can only have 1 struct per `mod` and that struct is declared outside the mod and has the same name as the mod?

**Comment by magnus-madsen** [comment]
> Ok, so basically, we can only have 1 struct per `mod` and that struct is declared outside the mod and has the same name as the mod?

You can still do:

```
mod Foo {
  struct A {}
  mod A {}
  struct B {}
  mod B{ }
}
```

If it helps, when you write:

```
struct S {
}
mod S {
  def foo
}
```

You can "almost" think of it like if Flix allowed members of structs:

```
struct S {
  //...
    def foo
}
```

**Comment by mlutze** [comment]
The key effect I see with this design is that there's no way to qualified-ly access struct members. This means that struct fields are all implicitly private and we'd have to make accessor/mutator functions to touch them outside the namespace. This is different from e.g. enums, where we can access the cases from outside the namespace by qualifying them.

I think this is OK, but I want to be sure this is the intention.

**Comment by dghosef** [comment]
I think practically that might just make people make getter and setter functions which becomes boilerplate.

It feels wrong for someone to be able to have access to a struct type but not be able to access its field. I.e. I feel that a structs fields should have the same publicity as the struct itself

**Comment by magnus-madsen** [comment]
> The key effect I see with this design is that there's no way to qualified-ly access struct members. This means that struct fields are all implicitly private and we'd have to make accessor/mutator functions to touch them outside the namespace. This is different from e.g. enums, where we can access the cases from outside the namespace by qualifying them.
> 
> I think this is OK, but I want to be sure this is the intention.

Yes this is correct.

**Comment by magnus-madsen** [comment]
> I think practically that might just make people make getter and setter functions which becomes boilerplate.
> 
> It feels wrong for someone to be able to have access to a struct type but not be able to access its field. I.e. I feel that a structs fields should have the same visibility as the struct itself

Partially agree, but this is a good place _to start_.

**Comment by magnus-madsen** [comment]
@dghosef It seems @sockmaster27 is blocked on this. Could I ask you to work on a fix with some urgency? Thanks! :)

**Comment by dghosef** [comment]
See my draft pr - I think it should be ready to go but requires the -> pr

**Comment by magnus-madsen** [comment]
> See my draft pr - I think it should be ready to go but requires the -> pr

Link?

**Comment by dghosef** [comment]
Sorry was eating lunch. https://github.com/flix/flix/pull/8411

**Comment by dghosef** [comment]
That one covers both -> and the mod thing

**Comment by JonathanStarup** [comment]
Fixed by #8411
